docs/language/informal/covariant-overrides.md

```diff
@@ -0,0 +1,742 @@
+widget.addChild(new Widget());
+```
+
+... does not. Both will fail at run time, but the whole point of allowing
+covariant overrides is that developers should be allowed to opt in and take
+that risk.
+
+#### Tear-offs
+
+The *static* type of a tear-off is the type declared in the statically known
+receiver type:
+
+```dart
+var closure = new RadioGroup().addChild;
+```
+
+Here, `closure` has static type `(RadioButton) -> void`.
+
+```dart
+var closure = (new RadioGroup() as Widget).addChild;
+```
+
+Here, it has static type `(Widget) -> void`.
+
+Note that in both cases, we're tearing off the same actual method at runtime. We
+discuss below which runtime type such a tear-off should have.
+
+### Runtime semantics
+
+The runtime semantics of the language with covariant overrides is the same
+as the runtime semantics of the language without that feature, except that
+some type checks are not guaranteed to succeed based on static checks, so
+they must be performed at run time.
+
+In particular, when a method with a covariant parameter is invoked, it is not
+possible to guarantee based on the static type check at call sites that
+the actual argument will have the type which is declared for that parameter
+in the method that is actually invoked. A dynamic type check must then be
+performed to verify that the actual argument is null or has that type,
+and a `CastError` must be thrown if the check fails.
+
+In other words, the static checks enforce the original type, and the dynamic
+check enforces the desired type, and both checks are required because the
+desired type can be a proper subtype of the original type. Here is an example:
+
+```dart
+Widget widget = new RadioGroup();
+try {
+  widget.addChild(new Widget());
+} on CastError {
+  print("Caught!"); // Gets printed.
+}
+```
+
+In practice, a compiler may generate code such that every covariant parameter
+in each method implementation gets checked before the body of the method is
+executed. This is a correct implementation strategy because the dynamic check
+is performed with the desired type and it is performed for all invocations,
+no matter which original type the call site had.
+
+As an optimization, the dynamic check could be omitted in some cases, because
+it is guaranteed to succeed. For instance, we may know statically that the
+original and the desired type are the same, because the receiver's type is
+known exactly:
+
+```dart
+new RadioGroup().add(myRadioButton())
+```
+
+#### The runtime type of a tear-off
+
+For any given declaration *D* of a method `m`, the reified type
+obtained by tearing off `m` from a receiver whose dynamic type is a
+class `C` wherein *D* is the most specific declaration of `m` (that
+is, *D* has not been overridden by any other method declaration in
+`C`) can be computed as follows: Collect the set of all declarations
+of `m` in direct and indirect supertypes of `C`. For the reified
+return type, choose the greatest lower bound of all declared return
```

Comment:
Just use the actual return type. It must be a subtype of all other return types, and there is nothing special about return types that makes you need to look at super declarations.
It's only because the "original type" is different from the "desired type" and "original type" isn't written in the function declaration, that we need to look at super-declarations to find it for covariant parameters.